addons/l10n_it_edi/models/res_partner.py

```diff
@@ -31,21 +32,135 @@ class ResPartner(models.Model):
             "Destination Code must have between 6 and 7 characters."),
     ]
 
-    @api.model
-    def _l10n_it_normalize_codice_fiscale(self, codice):
-        if codice and re.match(r'^IT[0-9]{11}$', codice):
-            return codice[2:13]
-        return codice
+    def _l10n_it_edi_get_values(self):
+        """ Generates all partner values needed by l10n_it_edi XML export """
+        if not self:
+            return False
+
+        self.ensure_one()
+        partner = self
+        class NormalizedPartnerValues(UserDict):
+            def __getitem__(self, attr):
+                """First lookup the the data dict and if miss then lookup on the partner """
+                if attr in self.data:
+                    return self.data[attr]
+                if attr in partner._fields:
+                    return partner[attr]
+                return False
+
+        return NormalizedPartnerValues({
+            'country_code': self._l10n_it_edi_normalized_country(),
+            'state_code': self._l10n_it_edi_normalize_state(),
+            'zip': self._l10n_it_edi_normalize_zip(),
+            'vat': self._l10n_it_edi_normalized_vat(),
+            'codice_fiscale': self._l10n_it_edi_normalized_codice_fiscale(),
+            'in_eu': self._l10n_it_edi_in_eu(),
+            'pa_index': self._l10n_it_edi_normalized_pa_index(),
+            'first_name': ' '.join(self.name.split()[:1]),
+            'last_name': ' '.join(self.name.split()[1:]),
+        })
+
+    def _l10n_it_edi_normalize_state(self):
+        self.ensure_one()
+        if self._l10n_it_edi_normalized_country() == 'IT' and self.state_id:
+            return self.state_id.code
+        else:
+            return False
+
+    def _l10n_it_edi_normalize_zip(self):
+        self.ensure_one()
+        if self._l10n_it_edi_normalized_country() == 'IT':
+            return self.zip
+        else:
+            return '00000'
+
+    def _l10n_it_edi_in_eu(self):
+        """ Determines whether the partner resides in Europe, useful for FatturaPA export """
+        self.ensure_one()
+        europe = self.env.ref('base.europe', raise_if_not_found=False)
+        country = self.country_id
+        return not europe or not country or country in europe.country_ids
+
+    def _l10n_it_edi_normalized_vat(self):
```

Comment:
_l10n_it_edi_normalized_vat and _l10n_it_edi_normalized_country seem to have the some logic and they are related to the same thing, so better to put it in the same method.

Reply:
I've grouped up all functions that didn't have a reason to be standalone.